source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -127,19 +148,24 @@ void SimpleHttpCache::updateHeaders(const LookupContext& lookup_context,
     return;
   }
 
-  // https://www.rfc-editor.org/rfc/pdfrfc/rfc7234.txt.pdf
-  // 4.3.4 Freshening Stored Responses upon Validation
-  // use other header fields provided in the 304 (Not Modified)
-  // response to replace all instances of the corresponding header
-  // fields in the stored response.
-  //
   // Assumptions:
   // 1. The internet is fast, i.e. we get the result as soon as the server sends it.
-  // Race conditions would not be possible because we are always processing up-to-date data.
+  //    Race conditions would not be possible because we are always processing up-to-date data.
   // 2. No key collision for etag. Therefore, if etag matches it's the same resource.
   // 3. Backend is correct. etag is being used as a unique identifier to the resource
-  // TODO(tangsaidi) merge the header map instead of replacing it according to rfc7234
-  entry.response_headers_ = Http::createHeaderMap<Http::ResponseHeaderMapImpl>(response_headers);
+
+  // use other header fields provided in the new response to replace all instances
+  // of the corresponding header fields in the stored response
+  response_headers.iterate(
+      [&entry](const Http::HeaderEntry& response_header) -> Http::HeaderMap::Iterate {
+        Http::LowerCaseString lower_case_key{response_header.key().getStringView()};
+        if (headersNotToUpdate.contains(lower_case_key)) {
+          return Http::HeaderMap::Iterate::Continue;
+        }
+        entry.response_headers_->remove(lower_case_key);
+        entry.response_headers_->addCopy(lower_case_key, response_header.value().getStringView());
```

Comment:
I'm assuming (though it's not 100% clear from context) that you are modifying a different set of headers than the one you are scanning. Is that right? Otherwise I think response_header.value() might not be valid after the remove. Maybe just comment about that?

In any case, rather than 'remove' and then 'addCopy' on the same header, can you just find the HeaderEntry and then set it? I think that would make the code a little more complex, but faster. It'll be more complex because you'd have to special-case when there was more than one header of the same name (which is likely less common).

See HeaderMap::get which returns a GetResult, and you can just set the value of that, if there's only one of them. Then you can do the remove/addCopy dance or something else only if there was more than one.

Reply:
Also I'm not sure if it is valid HTTP to duplicate any of the headers in your set. Of course you need to handle it, but it makes sense to optimize for the single-entry only case.

Reply:
> In any case, rather than 'remove' and then 'addCopy' on the same header, can you just find the HeaderEntry and then set it? I think that would make the code a little more complex, but faster. It'll be more complex because you'd have to special-case when there was more than one header of the same name (which is likely less common).

The `HeaderMap` doesn't have an API that returns a non-const reference to the entry of interest. Specifically, the `GetResult` class has the following documentation.
```
 /**
   * This is a wrapper for the return result from get(). It avoids a copy when translating from
   * non-const HeaderEntry to const HeaderEntry and only provides const access to the result.
   */
```

Adding this optimization means that we need to change the `header_map`'s API interface, which is used by lots of http filters. IMHO, the optimization is not worth the risk. It's still O(n) for the single-entry use case because we need to scan through the headermap to ensure there's no duplicate.

Reply:
OK I think you're right; there's no good way to do the multi-value get. But can you what happens if you use set? That might be a single-call mechanism that can give you the same semantics.

Just want to make sure that if there are multiple values that they'd all be replaced by the one value you are specifying.

Reply:
Thanks for pointing this out! It looks like `setCopy` does exactly remove and add. 
```